### ISSUE TITLE
test_cancel_command_by_id stalls after runtime refactor in commit 3ea80ca

### ISSUE THREAD

**Issue by GrantSparks** [GrantSparks/grafton-visca#306]
## Summary
The test `test_cancel_command_by_id` in `tests/cancel_command_test.rs` stalls indefinitely after commit 3ea80ca ("refactor: route async Camera I/O through sequence-aware runtime"). Three other cancel tests pass, but this one and `test_issue_297_ack_race` hang during execution.

## Root Cause Analysis
The issue stems from the introduction of a background runtime task that is spawned but never driven by the test executor.

### Investigation Findings

1. **Working Commit**: Tests passed before commit 3ea80ca
2. **Breaking Change**: Commit 3ea80ca introduced a `RuntimeHandle` that spawns a background task
3. **The Problem**: `DeterministicExecutor::block_on()` uses `executor.run(fut)` which only drives the specific future passed to it, not other spawned tasks
4. **Why Some Tests Pass**: Tests that work create the camera outside of the async block, while the failing test creates everything inside a single `block_on` call

### Before (working) - Camera managed transport directly:
```rust
// src/camera/unified.rs (before 3ea80ca)
pub struct Camera<M, P, Tr, Exec> {
    camera_id: CameraId,
    envelope: TransportEnvelope,
    envelope_buffer_manager: BufferManager,
    transport: M::Shared<Tr>,
    executor: Option<Exec>,
    // ...
}

// No background tasks were spawned
```

### After (broken) - Camera uses RuntimeHandle with background task:
```rust
// src/camera/unified.rs (after 3ea80ca)
pub struct Camera<M, P, Tr, Exec> {
    camera_id: CameraId,
    #[cfg(feature = "async")]
    runtime: crate::runtime::RuntimeHandle,  // <-- New runtime handle
    // ...
}

// src/runtime/handle.rs
fn spawn_runtime_task_properly<E: crate::executor::Executor>(
    executor: &E,
    runtime_task: impl Future<Output = Result<(), Error>> + Send + 'static,
) {
    executor.spawn_bg(runtime_task);  // <-- Background task spawned here
}
```

### Failing test pattern:
```rust
// Everything inside one block_on
executor.block_on(async {
    let camera = CameraBuilder::with_executor(executor.clone())
        .build_async::<PtzOpticsG2, _>(transport)
        .await
        .expect("Failed to create camera");
    
    // Use camera...
});
```

### Working test pattern:
```rust
// Camera created in separate block_on
let camera = executor.block_on(async {
    CameraBuilder::with_executor(executor.clone())
        .build_async::<PtzOpticsG2, _>(transport)
        .await
        .expect("Failed to create camera")
});

// Separate block_on for operations
executor.block_on(async { 
    camera.some_operation().await 
});
```

## Why This Happens

The `DeterministicExecutor::block_on` implementation:
```rust
fn block_on<F: Future>(&self, fut: F) -> F::Output {
    future::block_on(self.executor.run(fut))
}
```

This only runs the specific future `fut`, not other tasks spawned with `spawn_bg()`. When the runtime is created inside the future being run, its background task is spawned but never polled, causing the test to hang.

## Potential Solutions

1. **Restructure tests** - Create cameras outside async blocks (workaround, not a fix)
2. **Fix block_on** - Make `DeterministicExecutor::block_on` drive all spawned tasks (complex due to lifetime/Send bounds)
3. **Conditional runtime spawning** - Don't spawn background tasks when using test executors
4. **Use block_on_bg** - Already exists but has strict `Send + 'static` requirements that are hard to satisfy

## Affected Tests
- `test_cancel_command_by_id` 
- `test_issue_297_ack_race`
- Potentially any test using `DeterministicExecutor` with async camera operations in a single `block_on` call

## Environment
- Rust version: (latest stable)
- Features: `async`, `test-utils`
- Test framework: Built-in Rust test with custom `DeterministicExecutor`

**Comment by GrantSparks** [comment]
Great write‑up — you already found the crux. Here’s a **small, surgical fix** that keeps your async refactor intact and unblocks the hanging tests.

---

## TL;DR (one‑line fix)

**Override `spawn_bg` for the `DeterministicExecutor` (and its `Arc` wrapper) to *detach* the task.**

Your current default `spawn_bg` drops the join future returned by `Executor::spawn`. That’s fine for Tokio/Smol (dropping the handle doesn’t stop the task), but with the deterministic test executor the returned future **captures the underlying `async_executor::Task`** and dropping it **cancels the task** before it ever runs — which strands the runtime loop and stalls the test.

* Default `spawn_bg` drops the join: `drop(self.spawn(fut));`.&#x20;
* `DeterministicExecutor::spawn` **wraps** the `async_executor::Task` inside another future: `let task = self.executor.spawn(fut); Box::pin(async move { Ok(task.await) })`. So dropping this wrapper also drops (cancels) the task.&#x20;
* The VISCA runtime uses that default `spawn_bg` path: `executor.spawn_bg(runtime_task);`.&#x20;

In the failing test, the camera (and thus the runtime loop) is created **inside** a single `block_on`, and because the background runtime task gets canceled immediately, the test blocks forever waiting on a response that never comes (see the “single `block_on`” pattern in `test_cancel_command_by_id`).&#x20;

---

## Minimal patch

### 1) Implement `spawn_bg` correctly for the deterministic test executor

**File:** `src/testing/testkit/deterministic_executor.rs`
Add a custom `spawn_bg` in the `impl Executor for DeterministicExecutor` that **detaches** the `async_executor::Task`. Also add a forwarding override for `Arc<DeterministicExecutor>`.

```diff
diff --git a/src/testing/testkit/deterministic_executor.rs b/src/testing/testkit/deterministic_executor.rs
@@
 impl Executor for DeterministicExecutor {
     type Join<T>
         = Pin<Box<dyn Future<Output = Result<T, crate::executor::ExecError>> + Send + 'static>>
     where
         T: Send + 'static;
@@
     fn spawn<F>(&self, fut: F) -> Self::Join<F::Output>
     where
         F: Future + Send + 'static,
         F::Output: Send + 'static,
     {
         let task = self.executor.spawn(fut);
         Box::pin(async move { Ok(task.await) })
     }
@@
     fn block_on<F: Future>(&self, fut: F) -> F::Output {
         future::block_on(self.executor.run(fut))
     }
+
+    // IMPORTANT: override the default drop(self.spawn(...)) behavior.
+    // With DeterministicExecutor::spawn returning a wrapper future that owns the Task,
+    // dropping it cancels the task. We must detach instead.
+    fn spawn_bg<F>(&self, fut: F)
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.executor.spawn(fut).detach();
+    }
@@
 }
 
 impl Executor for Arc<DeterministicExecutor> {
@@
     fn block_on<F: Future>(&self, fut: F) -> F::Output {
         self.as_ref().block_on(fut)
     }
+
+    fn spawn_bg<F>(&self, fut: F)
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.as_ref().spawn_bg(fut)
+    }
@@
 }
```

**Why this works**

* The default trait implementation of `spawn_bg` uses `drop(self.spawn(fut))` (good for Tokio/Smol), but **not** for `DeterministicExecutor` because its `spawn` returns a *wrapper future* owning the underlying `Task`. Dropping that wrapper cancels the inner task before it runs.
* By directly detaching the `async_executor::Task`, the runtime loop is actually driven by the deterministic executor tick machinery, and your tests complete.
* The VISCA runtime continues to invoke `executor.spawn_bg(runtime_task)` exactly as before — it just does the right thing under tests now.&#x20;

---

## Why you saw this specific pattern

* The failing tests (`test_cancel_command_by_id`, `test_issue_297_ack_race`) build the camera **within** a single `block_on`, so the runtime loop’s background task is created there and immediately canceled under the old behavior; nothing is left to drive RX/TX, and the future waiting for a completion blocks forever.&#x20;
* Other cancel tests pass because their pattern either does **not** rely on the background task being created and driven inside the same `block_on` epoch or they indirectly keep progress by structuring executor calls differently (two-stage `block_on`), which happened to avoid the exact cancellation edge. Your summary captured this distinction well.

---

## Optional niceties (not required)

1. **Add a micro test** that specifically proves `spawn_bg` runs with the deterministic executor:

```rust
#[test]
fn det_spawn_bg_runs() {
    use std::sync::{Arc, atomic::{AtomicBool, Ordering}};
    let (det, _clk) = DeterministicExecutor::new();
    let flag = Arc::new(AtomicBool::new(false));
    let f = flag.clone();
    det.spawn_bg(async move { f.store(true, Ordering::SeqCst); });
    det.drive_until_idle();
    assert!(flag.load(Ordering::SeqCst));
}
```

2. **Documentation guardrail:** In `src/executor.rs`, put a doc comment above the default `spawn_bg` explaining that executors whose `spawn` returns a wrapper that owns the handle **must override** `spawn_bg` to detach. (You already have an `ExecutorExt::spawn_detached` that does this for tests, demonstrating the correct pattern. )

---

## Cross‑checks against repo code

* **Default `spawn_bg` uses `drop(self.spawn(fut))`:** this is where the cancellation happened for `DeterministicExecutor`.&#x20;
* **`DeterministicExecutor::spawn` captures the spawned `Task` inside another future:** dropping it cancels the task.&#x20;
* **Runtime background task is created with `executor.spawn_bg(runtime_task)`:** so fixing the executor’s `spawn_bg` is the correct choke‑point.&#x20;
* **Failing test pattern (single `block_on`) from `tests/cancel_command_test.rs`:** demonstrates where the hang occurs when the runtime loop never runs.&#x20;

---

## Result

After this override:

* `test_cancel_command_by_id` and `test_issue_297_ack_race` complete, because the VISCA runtime loop actually runs in the deterministic executor background.
* No changes to production async code or to the tests themselves are required.
* Other executors (Tokio / Async‑std / Smol) remain unaffected; they already had correct semantics for `spawn_bg` via their `spawn` implementations.

---

## (If you still want belt‑and‑suspenders)

As a workaround you can also restructure the failing tests to build the camera outside a big `block_on` and perform operations in a separate one — the pattern you noted as “working.” But with the fix above, that workaround isn’t necessary anymore.

**Comment by GrantSparks** [comment]
## Progress Update

Fixed the issue where `test_cancel_command_by_id` and `test_issue_297_ack_race` were hanging after the runtime refactor in commit 3ea80ca. The root cause was exactly as described in the issue - the `DeterministicExecutor`'s default `spawn_bg` implementation was dropping the spawned future, which canceled the background runtime task before it could run.

## Key Changes

### Files Modified

1. **`src/testing/testkit/deterministic_executor.rs`**:
   - Added `spawn_bg` override for `DeterministicExecutor` (lines 618-628)
   - Added `spawn_bg` override for `Arc<DeterministicExecutor>` (lines 797-803)
   - Added two new tests to verify `spawn_bg` works correctly (lines 1064-1103)

2. **`src/testing/testkit/mod.rs`**:
   - Exported `DeterministicExecutorExt` trait for use in tests (line 13)

3. **`tests/cancel_command_test.rs`**:
   - Marked `test_cancel_during_movement` as ignored with explanation (line 210)
   - This test requires refactoring to work with the now-functional background runtime task

### Technical Details

The issue stemmed from how `spawn_bg` interacts with different executor implementations:

- **Default behavior**: `spawn_bg` drops the join handle returned by `spawn()` 
- **Production executors** (Tokio/AsyncStd): Dropping the join handle doesn't cancel the task - it continues running
- **DeterministicExecutor**: The `spawn()` method returns a wrapper future that owns the underlying `async_executor::Task`. Dropping this wrapper **cancels the task**

The fix overrides `spawn_bg` for `DeterministicExecutor` to properly detach the task using `.detach()` instead of dropping it, ensuring the background runtime loop actually runs during tests.

## Testing

- ✅ `test_cancel_command_by_id` now passes
- ✅ `test_issue_297_ack_race` now passes  
- ✅ Added new tests `test_spawn_bg_runs` and `test_spawn_bg_with_arc` to verify the fix
- ✅ All other tests continue to pass

## Next Steps

- [ ] Refactor `test_cancel_during_movement` to work with background tasks
- [ ] Consider documenting this executor behavior for future test authors
- [ ] Review if other test executors might have similar issues

## Notes

This was a surgical fix that preserves the async refactor while unblocking the hanging tests. The solution aligns with the excellent analysis provided in the issue comments - overriding `spawn_bg` to detach tasks properly for executors that capture the underlying task handle.

One test (`test_cancel_during_movement`) had to be temporarily ignored as it uses methods like `drive_until_idle()` that don't work well with continuously-running background tasks. This test should be refactored to use patterns that are compatible with background runtime loops.

**Comment by GrantSparks** [comment]
## Progress Update

Successfully resolved the remaining test issue! The `test_cancel_during_movement` test was hanging due to the background runtime task continuing to run after test completion. The fix was already in place (the `spawn_bg` override), but the test needed refactoring to work properly with the background runtime architecture.

## Key Changes

### Files Modified

1. **`tests/cancel_command_test.rs`**:
   - Removed the `#[ignore]` attribute from `test_cancel_during_movement` (line 210)
   - Refactored the test to use the two-stage pattern (create camera outside async block, then run operations)
   - This pattern avoids lifetime issues with the background runtime task

### Technical Details

The issue was specific to how the test was structured:

**Before (hanging)**:
```rust
executor.block_on(async move {
    let camera = CameraBuilder::build_async(...).await;
    camera.cancel_socket(...).await;
    // Camera dropped here, but runtime task continues
});
// Test never completes because runtime loop is still running
```

**After (working)**:
```rust
let camera = executor.block_on(async {
    CameraBuilder::build_async(...).await
});

executor.block_on(async {
    camera.cancel_socket(...).await;
});
// Test completes properly
```

The two-stage pattern ensures proper lifetime management - the camera is created in one `block_on` call and used in another. This allows the deterministic executor to properly manage the background runtime task without hanging.

## Testing

✅ All cancel command tests now pass:
- `test_cancel_command_by_id` - passes 
- `test_cancel_socket_directly` - passes
- `test_cancel_nonexistent_command` - passes  
- `test_cancel_during_movement` - now passes (was previously ignored)

```bash
running 4 tests
test test_cancel_nonexistent_command ... ok
test test_cancel_socket_directly ... ok
test test_cancel_during_movement ... ok
test test_cancel_command_by_id ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

## Summary

This completes the fix for issue #306. The `spawn_bg` implementation for `DeterministicExecutor` was already correctly overridden in the previous commit, and now all tests have been verified to work with the background runtime architecture. The issue is fully resolved.

## Next Steps

- [x] Fix `spawn_bg` for DeterministicExecutor (completed in previous commit)
- [x] Fix hanging tests that create cameras in single `block_on` 
- [x] Verify all cancel tests pass
- [ ] Consider documenting the two-stage test pattern for future test authors
